br/pkg/lightning/restore/check_info.go

```diff
@@ -658,86 +660,260 @@ func (rc *Controller) SchemaIsValid(ctx context.Context, tableInfo *mydump.MDTab
+			colMap[col.Name.L] = struct{}{}
+		}
+	}
+	// tidb_rowid can be ignored in check
+	colMap[model.ExtraHandleName.String()] = struct{}{}
+	for _, col := range colsFromDataFile {
+		if _, ok := colMap[col]; !ok {
+			checkMsg := "please check table schema"
+			if dataFileMeta.Type == mydump.SourceTypeCSV && rc.cfg.Mydumper.CSV.Header {
+				checkMsg += " and csv file header"
 			}
-			for _, col := range core.Columns {
-				if _, ok := colMap[col.Name.L]; ok {
-					// tidb's column is ignored
-					// we need ensure this column has the default value.
-					if _, hasDefault := defaultCols[col.Name.L]; !hasDefault {
-						msgs = append(msgs, fmt.Sprintf("TiDB schema `%s`.`%s`'s column %s cannot be ignored,"+
-							"because it doesn't hava a default value, please set tables.ignoreColumns properly",
-							tableInfo.DB, tableInfo.Name, col.Name.L))
-					}
-				} else {
-					colMap[col.Name.L] = struct{}{}
-				}
+			msgs = append(msgs, fmt.Sprintf("TiDB schema `%s`.`%s` doesn't have column %s, "+
+				"%s or use tables.ignoreColumns to ignore %s",
+				tableInfo.DB, tableInfo.Name, col, checkMsg, col))
+		} else {
+			// remove column for next iteration
+			delete(colMap, col)
+		}
+	}
+	// if theses rest columns don't have a default value.
+	for col := range colMap {
+		if _, ok := defaultCols[col]; ok {
+			continue
+		}
+		msgs = append(msgs, fmt.Sprintf("TiDB schema `%s`.`%s` doesn't have the default value for %s"+
+			"please give a default value for %s or choose another column to ignore or add this column in data file",
+			tableInfo.DB, tableInfo.Name, col, col))
+	}
+	return msgs, nil
+}
+
+// checkCSVHeader try to check whether the csv header config is consistent with the source csv files
+func (rc *Controller) checkCSVHeader(ctx context.Context, dbMetas []*mydump.MDDatabaseMeta) error {
+	// if cfg set header = ture but source files actually contain not header, former SchemaCheck should
+	// return error in this situation, so we need do it again.
+	if rc.cfg.Mydumper.CSV.Header {
+		return nil
+	}
+	var (
+		tableMeta    *mydump.MDTableMeta
+		csvCount     int
+		hasUniqueIdx bool
+	)
+	// only check one table source files for better performance. The checked table is choosed based on following two factor:
+	// 1. contains at least 1 csv source file, 2 is preferable
+	// 2. table schema contains primary key or unique key
+	// if the two factors can't be both satisfied, the first one has a higher priority
+outer:
+	for _, dbMeta := range dbMetas {
+		for _, tblMeta := range dbMeta.Tables {
+			if len(tblMeta.DataFiles) == 0 {
+				continue
 			}
-			// tidb_rowid can be ignored in check
-			colMap[model.ExtraHandleName.String()] = struct{}{}
-			for _, col := range colsFromDataFile {
-				if _, ok := colMap[col]; !ok {
-					checkMsg := "please check table schema"
-					if dataFileMeta.Type == mydump.SourceTypeCSV && rc.cfg.Mydumper.CSV.Header {
-						checkMsg += " and csv file header"
+			tableHasUniqueIdx := false
+			csvCnt := 0
```

Comment:
```suggestion
			tableCSVCount := 0
```

it is hard to understand what is the difference of `csvCnt` and `csvCount`.

Reply:
😅 naming is the one of two difficulties in computer science.